src/ring_theory/perfection.lean

```diff
@@ -0,0 +1,366 @@
+/-
+Copyright (c) 2020 Kenny Lau. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Kenny Lau
+-/
+
+import algebra.char_p
+import algebra.ring.pi
+import analysis.special_functions.pow
+import field_theory.perfect_closure
+import ring_theory.localization
+import ring_theory.subring
+import ring_theory.valuation.integers
+
+/-!
+# Ring Perfection and Tilt
+
+In this file we define the perfection of a ring of characteristic p, and the tilt of a field
+given a valuation to `ℝ≥0`.
+
+## TODO
+
+Define the valuation on the tilt, and define a characteristic predicate for the tilt.
+
+-/
+
+universes u v w
+
+open_locale nnreal
+
+/-- The perfection of a monoid `M`, defined to be the projective limit of `M`
+using the `p`-th power maps `M → M` indexed by the natural numbers, implemented as
+`{ f : ℕ → M | ∀ n, f (n + 1) ^ p = f n }`. -/
+def monoid.perfection (M : Type u) [comm_monoid M] (p : ℕ) : submonoid (ℕ → M) :=
+{ carrier := { f | ∀ n, f (n + 1) ^ p = f n },
+  one_mem' := λ n, one_pow _,
+  mul_mem' := λ f g hf hg n, (mul_pow _ _ _).trans $ congr_arg2 _ (hf n) (hg n) }
+
+/-- The perfection of a semiring `R` with characteristic `p`,
+defined to be the projective limit of `R` using the Frobenius maps `R → R`
+indexed by the natural numbers, implemented as `{ f : ℕ → R | ∀ n, f (n + 1) ^ p = f n }`. -/
+def comm_semiring.perfection (R : Type u) [comm_semiring R]
```

Comment:
```suggestion
def semiring.perfection (R : Type u) [comm_semiring R]
```

Reply:
Other idea, but I'm not wed to it: should we change all the names to `perfection.submonoid`, `perfection.subsemiring`, etc?

Reply:
Then they would be 3 characters longer, which in my opinion is not the best scenario.